source/extensions/filters/network/thrift_proxy/conn_manager.cc

```diff
@@ -220,6 +222,30 @@ FilterStatus ConnectionManager::ResponseDecoder::messageBegin(MessageMetadataSha
   if (metadata->hasReplyType()) {
     success_ = metadata->replyType() == ReplyType::Success;
   }
+
+  // Check if the upstream host is draining.
+  //
+  // Note: the drain header needs to be checked here in messageBegin, and not transportBegin, so
+  // that we can support the header in TTwitter protocol, which reads/adds response headers to
+  // metadata in messageBegin when reading the response from upstream. Therefore detecting a drain
+  // should happen here.
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.thrift_connection_draining")) {
+    metadata_->setDraining(!metadata->headers().get(Headers::get().Drain).empty());
+    metadata->headers().remove(Headers::get().Drain);
+  }
+
+  // Check if this host itself is draining.
+  //
+  // Note: Similarly as above, the response is buffered until transportEnd. Therefore metadata
+  // should be set before the encodeFrame() call. It should be set at or after the messageBegin call
+  // so that the header is added after all upstream headers passed, due to messageBegin possibly not
+  // getting headers in transportBegin.
+  ConnectionManager& cm = parent_.parent_;
+  if (cm.drain_decision_.drainClose()) {
+    // TODO(rgs1): should the key value contain something useful? E.g.: minutes til drain is over?
+    metadata->headers().addReferenceKey(Headers::get().Drain, "true");
```

Comment:
Should this logic be gated by the runtime guard as well?

Reply:
My thinking was no, since it only adds the header. So for example, for rollout it could be nice to see the header even if you are not acting on it (yet). We could also have a stat incremented here now that I come to think of it.

But I am happy to gate this part too if that feels more consistent.

cc: @fishcakez

Reply:
discussed offline with @fishcakez – yeah let's put the header adding part besides the runtime guard too. And, let's add a stat - `downstream_resp_drain_close` - while we are add it.